Modules/_threadmodule.c

```diff
@@ -1435,6 +1435,7 @@ local_new(PyTypeObject *type, PyObject *args, PyObject *kw)
     Py_DECREF(localsdict);
     Py_DECREF(sentinel_wr);
 
+    _PyObject_SetDeferredRefcount((PyObject *)self);
```

Comment:
I think this needs to come before the call to `create_localsdict` above. `create_localsdict` creates a weakref to `self`, which causes this assert to fail in debug builds:
https://github.com/python/cpython/blob/087bb48acac997c06e69dae25bae2dd75194b980/Objects/object.c#L2557